File: ts/src/db/archive.ts
// SPDX-License-Identifier: Apache-2.0
import { join } from "node:path";
import { statSync, existsSync } from "node:fs";
import { Glob } from "bun";
import type { Tweet } from "../core/parse";
import { Store } from "./store";

/** Locate the tweets file inside an archive directory or accept a direct path. */
export function resolveTweetsFile(input: string): string | null {
  if (!existsSync(input)) return null;

  try {
    const stat = statSync(input);
    if (stat.isFile()) {
      return input;
    }

    const relatives = ["data/tweets.js", "data/tweet.js", "tweets.js", "tweet.js"];
    for (const rel of relatives) {
      const candidate = join(input, rel);
      if (existsSync(candidate) && statSync(candidate).isFile()) {
        return candidate;
      }
    }

    // Try a recursive glob scan for robustness
    const glob = new Glob("**/{tweets,tweet}.js");
    for (const file of glob.scanSync(input)) {
      const full = join(input, file);
      if (statSync(full).isFile()) {
        return full;
      }
    }
  } catch {
    // ignore
  }

  return null;
}

/** Strip the window.YTD.* assignment prefix and parse the JSON array. */
export function parseArchiveArray(raw: string): any[] {
  const start = raw.indexOf("[");
  const end = raw.lastIndexOf("]");
  if (start === -1 || end === -1 || end < start) {
    throw new Error("Archive file has no JSON array or delimiters are inverted");
  }
  const slice = raw.substring(start, end + 1);
  return JSON.parse(slice);
}

function countField(legacy: any, key: string): number {
  const val = legacy?.[key];
  if (typeof val === "number") return val;
  if (typeof val === "string") {
    const num = parseInt(val, 10);
    return isNaN(num) ? 0 : num;
  }
  return 0;
}

/** Convert one archive element into a Tweet structure. */
export function archiveTweetToTweet(elem: any, owner: { username: string; name: string }): Tweet | null {
  const legacy = elem.tweet || elem;
  const id = legacy.id_str;
  if (!id) return null;

  const text = legacy.full_text || legacy.text || "";

  return {
    id,
    text,
    author: owner,
    author_id: undefined,
    created_at: legacy.created_at || undefined,
    reply_count: countField(legacy, "reply_count"),
    retweet_count: countField(legacy, "retweet_count"),
    like_count: countField(legacy, "favorite_count") || countField(legacy, "like_count"),
    quote_count: countField(legacy, "quote_count"),
    view_count: undefined,
    conversation_id: legacy.conversation_id_str || undefined,
    in_reply_to_status_id: legacy.in_reply_to_status_id_str || undefined,
    lang: legacy.lang || undefined,
    is_note_tweet: false,
    quoted_tweet: null,
  };
}

/** Import a tweets archive into the SQLite store. */
export async function importArchive(
  store: Store,
  path: string,
  ownerHandle: string
): Promise<number> {
  const file = resolveTweetsFile(path);
  if (!file) {
    throw new Error(`No tweets.js or tweet.js found at ${path}`);
  }

  // Use high-performance native Bun.file reader
  const raw = await Bun.file(file).text();
  const arr = parseArchiveArray(raw);

  const cleanHandle = ownerHandle.replace(/^@/, "");
  const owner = {
    username: cleanHandle,
    name: cleanHandle,
  };

  let imported = 0;
  for (const elem of arr) {
    const tweet = archiveTweetToTweet(elem, owner);
    if (tweet) {
      store.upsertTweet(tweet);
      store.addEdge(owner.username, "authored", tweet.id);
      imported++;
    }
  }

  return imported;
}

File: src/archive.rs
// SPDX-License-Identifier: Apache-2.0
//! Twitter/X data-archive import into the local store.
//!
//! A Twitter data export ships `data/tweets.js` (and `tweet.js` on older
//! exports) as a JS-assignment-wrapped JSON array:
//!
//! ```text
//! window.YTD.tweets.part0 = [ { "tweet": { "id_str": ..., "full_text": ... } }, ... ]
//! ```
//!
//! This module strips the assignment prefix, parses the array, maps each
//! legacy tweet object onto our canonical [`Tweet`], and upserts it into the
//! store with an `authored` edge — the same shape live sync produces, so an
//! archive and live data merge seamlessly.

use std::path::{Path, PathBuf};

use serde_json::Value;

use crate::parse::{Author, Tweet};
use crate::store::{edge, Store};
use crate::{Result, XError};

/// Locate the tweets file inside an archive: accept a direct file, or a
/// directory containing `data/tweets.js` / `data/tweet.js` / `tweets.js`.
fn resolve_tweets_file(input: &Path) -> Option<PathBuf> {
    if input.is_file() {
        return Some(input.to_path_buf());
    }
    for rel in ["data/tweets.js", "data/tweet.js", "tweets.js", "tweet.js"] {
        let candidate = input.join(rel);
        if candidate.is_file() {
            return Some(candidate);
        }
    }
    None
}

/// Strip the `window.YTD.* = ` assignment prefix and parse the JSON array.
fn parse_archive_array(raw: &str) -> Result<Vec<Value>> {
    let start = raw
        .find('[')
        .ok_or_else(|| XError::Auth("archive file has no JSON array".into()))?;
    let end = raw
        .rfind(']')
        .ok_or_else(|| XError::Auth("archive file has no JSON array end".into()))?;
    if end < start {
        return Err(XError::Auth("archive file array delimiters inverted".into()));
    }
    let arr: Vec<Value> = serde_json::from_str(&raw[start..=end])?;
    Ok(arr)
}

fn count_field(legacy: &Value, key: &str) -> u64 {
    legacy
        .get(key)
        .and_then(|v| v.as_str().and_then(|s| s.parse().ok()).or_else(|| v.as_u64()))
        .unwrap_or(0)
}

/// Convert one archive `{ "tweet": {legacy} }` element into a [`Tweet`].
fn archive_tweet_to_tweet(elem: &Value, owner: &Author) -> Option<Tweet> {
    let legacy = elem.get("tweet").unwrap_or(elem);
    let id = legacy.get("id_str").and_then(Value::as_str)?.to_owned();
    let text = legacy
        .get("full_text")
        .or_else(|| legacy.get("text"))
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_owned();
    Some(Tweet {
        id: id.clone(),
        text,
        author: owner.clone(),
        author_id: None,
        created_at: legacy
            .get("created_at")
            .and_then(Value::as_str)
            .map(ToOwned::to_owned),
        reply_count: count_field(legacy, "reply_count"),
        retweet_count: count_field(legacy, "retweet_count"),
        like_count: count_field(legacy, "favorite_count"),
        quote_count: count_field(legacy, "quote_count"),
        view_count: None,
        conversation_id: legacy
            .get("conversation_id_str")
            .and_then(Value::as_str)
            .map(ToOwned::to_owned),
        in_reply_to_status_id: legacy
            .get("in_reply_to_status_id_str")
            .and_then(Value::as_str)
            .map(ToOwned::to_owned),
        lang: legacy
            .get("lang")
            .and_then(Value::as_str)
            .map(ToOwned::to_owned),
        is_note_tweet: false,
        quoted_tweet: None,
    })
}

/// Import an archive into `store`, attributing tweets to `owner_handle`.
///
/// Returns the number of tweets imported.
pub fn import_archive(store: &Store, path: &Path, owner_handle: &str) -> Result<usize> {
    let file = resolve_tweets_file(path).ok_or_else(|| {
        XError::Auth(format!(
            "no tweets.js found at {} (expected a tweets.js file or an archive dir)",
            path.display()
        ))
    })?;
    let raw = std::fs::read_to_string(&file)?;
    let arr = parse_archive_array(&raw)?;

    let owner = Author {
        username: owner_handle.trim_start_matches('@').to_owned(),
        name: owner_handle.trim_start_matches('@').to_owned(),
    };

    let mut imported = 0usize;
    for elem in &arr {
        if let Some(tweet) = archive_tweet_to_tweet(elem, &owner) {
            store.upsert_tweet(&tweet)?;
            store.add_edge(&owner.username, edge::AUTHORED, &tweet.id)?;
            imported += 1;
        }
    }
    Ok(imported)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_js_wrapped_archive() {
        let raw = r#"window.YTD.tweets.part0 = [
          { "tweet" : { "id_str": "1", "full_text": "hello archive", "favorite_count": "12", "created_at": "Wed May 22 10:00:00 +0000 2026" } },
          { "tweet" : { "id_str": "2", "full_text": "second", "retweet_count": "3" } }
        ]"#;
        let arr = parse_archive_array(raw).unwrap();
        assert_eq!(arr.len(), 2);
        let owner = Author {
            username: "aphrody_code".into(),
            name: "aphrody_code".into(),
        };
        let t = archive_tweet_to_tweet(&arr[0], &owner).unwrap();
        assert_eq!(t.id, "1");
        assert_eq!(t.text, "hello archive");
        assert_eq!(t.like_count, 12);
        assert_eq!(t.author.username, "aphrody_code");
        let t2 = archive_tweet_to_tweet(&arr[1], &owner).unwrap();
        assert_eq!(t2.retweet_count, 3);
    }

    #[test]
    fn handles_plain_json_array() {
        let raw = r#"[ { "tweet": { "id_str": "9", "full_text": "x" } } ]"#;
        let arr = parse_archive_array(raw).unwrap();
        assert_eq!(arr.len(), 1);
    }

    #[test]
    fn rejects_non_array() {
        assert!(parse_archive_array("window.YTD.tweets.part0 = {}").is_err());
    }
}
